internal/catalog/internal/types/workload.go

```diff
@@ -27,6 +28,118 @@ func RegisterWorkload(r resource.Registry) {
 	r.Register(resource.Registration{
 		Type:     WorkloadV1Alpha1Type,
 		Proto:    &pbcatalog.Workload{},
-		Validate: nil,
+		Validate: ValidateWorkload,
 	})
 }
+
+func ValidateWorkload(res *pbresource.Resource) error {
+	var workload pbcatalog.Workload
+
+	if err := res.Data.UnmarshalTo(&workload); err != nil {
+		return newErrDataParse(&workload, err)
+	}
+
+	var err error
+	// Only validate the identity if its not empty, if it is
+	// empty it will be defaulted elsewhere.
+	if workload.Identity != "" {
+		if !isValidDNSLabel(workload.Identity) {
+			err = multierror.Append(err, ErrInvalidField{
+				Name:    "identity",
+				Wrapped: errNotDNSLabel,
+			})
+		}
+	}
+
+	// Node associations are optional but if present the name should
+	// be a valid DNS label.
+	if workload.NodeName != "" {
+		if !isValidDNSLabel(workload.NodeName) {
+			err = multierror.Append(err, ErrInvalidField{
+				Name:    "node_name",
+				Wrapped: errNotDNSLabel,
+			})
+		}
+	}
+
+	if len(workload.Addresses) < 1 {
+		err = multierror.Append(err, ErrInvalidField{
+			Name:    "addresses",
+			Wrapped: errEmpty,
+		})
+	}
+
+	// Validate Workload Addresses
+	for idx, addr := range workload.Addresses {
+		if addrErr := validateWorkloadAddress(addr, workload.Ports); addrErr != nil {
+			err = multierror.Append(err, ErrInvalidListElement{
+				Name:    "addresses",
+				Index:   idx,
+				Wrapped: addrErr,
+			})
+		}
+	}
+
+	// Validate that the workload has at least one port
+	if len(workload.Ports) < 1 {
+		err = multierror.Append(err, ErrInvalidField{
+			Name:    "ports",
+			Wrapped: errEmpty,
+		})
+	}
+
+	// Validate the Workload Ports
+	for portName, port := range workload.Ports {
+		if portNameErr := validatePortName(portName); portNameErr != nil {
+			err = multierror.Append(err, ErrInvalidMapKey{
+				Map:     "ports",
+				Key:     portName,
+				Wrapped: portNameErr,
+			})
+		}
+
+		// disallow port 0 for now
+		if port.Port < 1 || port.Port > 65535 {
+			err = multierror.Append(err, ErrInvalidMapValue{
+				Map: "ports",
+				Key: portName,
+				Wrapped: ErrInvalidField{
+					Name:    "port",
+					Wrapped: errInvalidPhysicalPort,
+				},
+			})
+		}
+	}
```

Comment:
Should we validate that if there's a port with a mesh protocol, then there should be only one of these?

Reply:
Yes, I think so. This is now implemented.